Libraries/LibTLS/TLSv12.cpp

```diff
@@ -0,0 +1,516 @@
+{
+    OIDChain local_chain;
+    local_chain.root = chain;
+    size_t position = 0;
+
+    // parse DER...again
+    size_t index = 0;
+    u8 oid[16] { 0 };
+
+    local_chain.oid = oid;
+    if (has_key)
+        *has_key = 0;
+
+    u8 local_has_key = 0;
+    const u8* cert_data = nullptr;
+    size_t cert_length = 0;
+    while (position < size) {
+        size_t start_position = position;
+        if (size - position < 2) {
+            dbg() << "not enough data for certificate size";
+            return (i8)Error::NeedMoreData;
+        }
+        u8 first = buffer[position++];
+        u8 type = first & 0x1f;
+        u8 constructed = first & 0x20;
+        size_t octets = 0;
+        u32 temp;
+        index++;
+
+        if (level <= 0xff)
+            fields[level - 1] = index;
+
+        size_t length = _get_asn1_length((const u8*)&buffer[position], size - position, octets);
+
+        if (octets > 4 || octets > size - position) {
+            dbg() << "could not read the certificate";
+            return position;
+        }
+
+        position += octets;
+        if (size - position < length) {
+            dbg() << "not enough data for sequence";
+            return (i8)Error::NeedMoreData;
+        }
+
+        if (length && constructed) {
+            switch (type) {
+            case 0x03:
+                break;
+            case 0x10:
+                if (level == 2 && index == 1) {
+                    cert_length = length + position - start_position;
+                    cert_data = buffer + start_position;
+                }
+                // public key data
+                if (!cert.version && _asn1_is_field_present(fields, Constants::priv_der_id)) {
+                    temp = length + position - start_position;
+                    if (cert.der.size() < temp) {
+                        cert.der.grow(temp);
+                    } else {
+                        cert.der.trim(temp);
+                    }
+                    cert.der.overwrite(0, buffer + start_position, temp);
+                }
+                break;
+
+            default:
+                break;
+            }
+            local_has_key = false;
+            _parse_asn1(context, cert, buffer + position, length, level + 1, fields, &local_has_key, client_cert, root_oid, &local_chain);
+            if ((local_has_key && (!context.is_server || client_cert)) || (client_cert || _asn1_is_field_present(fields, Constants::pk_id))) {
+                temp = length + position - start_position;
+                if (cert.der.size() < temp) {
+                    cert.der.grow(temp);
+                } else {
+                    cert.der.trim(temp);
+                }
+                cert.der.overwrite(0, buffer + start_position, temp);
+            }
+        } else {
+            switch (type) {
+            case 0x00:
```

Comment:
Styling, feel free to ignore : Same remark about magic values here, what are they used for / can they be replaced with `#DEFINE`, constexpr or enums ?